### PR TITLE
First cut working towards a public API

### PR DESCRIPTION
**Pull request by mhammond**
[obviously *not* a real PR - I'm abusing this as a way to annotate the diff]

A first cut which should be considered a POC - although it's a *working* POC - all xpcshell tests appear to work.

High-level notes:
* There should be no significant functionality changes. Some things are spelled differently, and there are one or 2 edge-cases where a different exception is thrown than was before (specifically, when we decide to auto logout the user for reasons). There's also a change where we now send the "onverified" notification before we've actually fetched keys, both to keep things cleaner and to break assumptions that all logins must have keys available.

* This tries to remove some of the `internal` strangeness in `FxAccounts.jsm` and replaces it with a (somewhat tedious) `_internal` object - but at least now it is a *normal* object and not magic. There's no more "freeze".

* This `_internal` object is, err, internal - but it's internal to the entire `services/fxaccounts` structure - ie, FxAccountsPush.jsm et al all are considered free to use this _internal object. Indeed, there's a case to be made they should *exclusively* use this `_internal` object - indeed `FxAccountsPush.jsm` does use only the `_internal`, and all other modules should possibly do the same (although more refactoring will be necessary to make this work.

* Most of the "keys" magic has been moved from `FxAccounts.jsm` to a new `FxAccountsKeys.jsm`

* There are lots of "XXX" comments remaining, but most of these should be shallow.

I'll add further annotations in the patch itself.

### CODE REVIEW

File: services/fxaccounts/FxAccountsCommands.js [hunk 1]
```diff
@@ -110,7 +108,7 @@ class FxAccountsCommands {
     if (!sessionToken) {
       throw new Error("No session token.");
     }
-    const client = this._fxAccounts.getAccountsClient();
+    const client = this._fxAccounts._internal.fxAccountsClient;
```
Comment (mhammond) [inline services/fxaccounts/FxAccountsCommands.js]: This usage of `_internal` has crept into most of the "sub-modules" and also into tests. I'm hoping to experiment with making the internal object the only API used internally, but I'm not sure how that's going to work out. It might end up slightly cleaner to have some of them use 2 objects - `this._fxAccounts` and (say) `this._fxai`?

File: services/fxaccounts/FxAccountsPush.jsm [hunk 2]
```diff
@@ -51,9 +51,9 @@ FxAccountsPushService.prototype = {
    */
   pushService: null,
   /**
-   * Instance of FxAccounts or a mocked object.
+   * Instance of FxAccountsInternal or a mocked object.
    */
-  fxAccounts: null,
+  fxai: null,
```
Comment (mhammond) [inline services/fxaccounts/FxAccountsPush.jsm]: This is the only of the sub-modules which moved to using the internal object exclusively.

File: services/fxaccounts/tests/xpcshell/test_accounts.js [hunk 3]
```diff
@@ -180,6 +180,12 @@ function MockFxAccounts() {
     fxAccountsClient: new MockFxAccountsClient(),
     observerPreloads: [],
   });
+  // and for convenience so we don't have to touch as many lines in this test :)
+  // XXX - fix me - just touch those lines!
+  result.setSignedInUser = function(creds) {
```
Comment (mhammond) [inline services/fxaccounts/tests/xpcshell/test_accounts.js]: `setSignedInUser` isn't part of the public API, so I should update this test to touch even more lines.

File: services/fxaccounts/tests/xpcshell/test_accounts.js [hunk 4]
```diff
@@ -274,7 +287,7 @@ add_task(async function test_set_signed_in_user_signs_out_previous_account() {
   };
   await account.setSignedInUser(credentials);
 
-  account.internal._signOutServer = () => {
+  account._internal._signOutServer = () => {
```
Comment (mhammond) [inline services/fxaccounts/tests/xpcshell/test_accounts.js]: sooo many `._internal` additions everywhere :(

File: services/fxaccounts/tests/xpcshell/test_accounts.js [hunk 5]
```diff
@@ -802,16 +815,21 @@ add_task(async function test_getKeys_nonexistent_account() {
   });
 
   try {
-    await fxa.internal.getKeys();
+    await fxa.keys.getKeys();
     Assert.ok(false);
   } catch (err) {
-    Assert.equal(err.code, 401);
-    Assert.equal(err.errno, ERRNO_INVALID_AUTH_TOKEN);
+    // XXX - a recent refactor means that this now throws with the exception
```
Comment (mhammond) [inline services/fxaccounts/tests/xpcshell/test_accounts.js]: by "a recent refactor" I mean "this refactor" - specifically, we use `this._withCurrentAccountState` in places where we didn't before, and it is that which is throwing the new exception.

File: services/fxaccounts/tests/xpcshell/test_accounts.js [hunk 6]
```diff
@@ -919,31 +937,32 @@ add_task(async function test_getScopedKeys_unavailable_key() {
...
       `${SCOPE_OLD_SYNC} otherkeybearingscope profile`,
       "123456789a"
     ),
     /Unavailable key material for otherkeybearingscope/
   );
 });
 
-//  fetchAndUnwrapKeys with no keyFetchToken should trigger signOut
+// fetchAndUnwrapKeys with no keyFetchToken should trigger signOut
+// XXX - actually, it probably shouldn't in a decoupled world.
```
Comment (mhammond) [inline services/fxaccounts/tests/xpcshell/test_accounts.js]: Yeah, no idea why it does this - it should maybe just enter a "needs reauth" state? However, so long as `fxa.keys.canFetchKeys()` returns false in this case and doesn't force a signout we are probably OK.

File: services/sync/modules/browserid_identity.js [hunk 7]
```diff
@@ -324,11 +324,11 @@ this.BrowserIDManager.prototype = {
    * Provide override point for testing token expiration.
    */
   _now() {
-    return this._fxaService.now();
+    return this._fxaService._internal.now();
```
Comment (mhammond) [inline services/sync/modules/browserid_identity.js]: The use of some `._internal` stuff here is unfortunate, but in reality, this module really is a "peer" of the fxa modules, and most of the `_internal` stuff is for testing. It would still be nice to clean more of this up.

File: services/sync/modules/browserid_identity.js [hunk 8]
```diff
@@ -510,7 +510,7 @@ this.BrowserIDManager.prototype = {
     let getToken = async () => {
```
Comment (mhammond) [inline services/sync/modules/browserid_identity.js]: I think this `getToken` method could probably be moved to the new FxAccountsKeys.jsm?

File: services/fxaccounts/FxAccounts.jsm [hunk 9]
```diff
@@ -392,12 +358,442 @@ function urlsafeBase64Encode(key) {
...
+ * XXX - TODO - *all* non-underscore stuff here should have sphinx docstrings so
+ * that docs magically appear on https://firefox-source-docs.mozilla.org/.
  */
-var FxAccounts = function(mockInternal) {
-  let external = {};
-  let internal;
+class FxAccounts {
+  constructor(mocks = null) {
+    this._internal = new FxAccountsInternal();
+    this.keys = new FxAccountsKeys(this);
```
Comment (mhammond) [inline services/fxaccounts/FxAccounts.jsm]: `keys` could probably be a getter with the real object living on `._internal` like `config` is below. This makes `.keys` etc part of the public API, but I don't think we can avoid that - this entire patch model still implies we trust the consumers of the API to do the right thing, and that includes only trying to use/get keys when appropriate.

File: services/fxaccounts/FxAccountsCommands.js [hunk 10]
```diff
@@ -284,17 +282,15 @@ class SendTab {
...
-        await updateUserData({
-          device: {
-            ...device,
-            sendTabKeys,
-          },
-        });
-      }
-    );
+    await this._fxAccounts._withCurrentAccountState(async state => {
+      const { device } = await state.getUserAccountData();
```
Comment (rfk) [inline services/fxaccounts/FxAccountsCommands.js]: `getUserAccountData(["device"])` here?

File: services/fxaccounts/FxAccountsKeys.jsm [hunk 11]
```diff
@@ -0,0 +1,331 @@
...
+
+  fetchAndUnwrapKeys(keyFetchToken) {
+    return this.fxa._withCurrentAccountState(async currentState => {
+      if (logPII) {
+        log.debug("fetchAndUnwrapKeys: token: " + keyFetchToken);
+      }
+      // Sign out if we don't have a key fetch token.
+      if (!keyFetchToken) {
+        log.warn("improper fetchAndUnwrapKeys() call: token missing");
+        await this.fxa.signOut();
```
Comment (rfk) [inline services/fxaccounts/FxAccountsKeys.jsm]: This seem pretty brutal, I wonder if we could go to some sort of "repair connection" state rather than signing all the way out.

File: services/sync/modules/browserid_identity.js (hunk 7 above)
Comment (rfk) [reply services/sync/modules/browserid_identity.js]: Also we would like this module to go away at some point and be replaced by OAuth stuff, so I'm fine with messing it up a little in the meantime.

File: services/fxaccounts/FxAccounts.jsm [hunk 12]
```diff
@@ -392,12 +358,442 @@ function urlsafeBase64Encode(key) {
...
+  }
+
+  getDeviceId() {
+    return this._withCurrentAccountState(async currentAccountState => {
+      let data = await currentAccountState.getUserAccountData();
+      if (!data) {
+        // Without a signed-in user, there can be no device id.
+        return null;
+      }
+      // Try migrating first. Remove this in Firefox 65+.
```
Comment (rfk) [inline services/fxaccounts/FxAccounts.jsm]: Can in we in fact remove this, as the comment promises? :-)

File: services/fxaccounts/FxAccounts.jsm [hunk 13]
```diff
@@ -392,12 +358,442 @@ function urlsafeBase64Encode(key) {
...
+        // is signed in but this pref isn't set, things go weird.
+        // However, some thing do unconditionally load fxaccounts, such as
+        // about:prefs. When that happens we can detect the state and re-add the
+        // pref. Note that we only do this for verified users as that's what sync
+        // does (ie, if the user is unverified, sync will set it on verification)
+        if (
+          !Services.prefs.prefHasUserValue("services.sync.username") &&
+          data.email
+        ) {
+          Services.prefs.setStringPref("services.sync.username", data.email);
```
Comment (rfk) [inline services/fxaccounts/FxAccounts.jsm]: I guess we change this somehow in the face of the decoupling work, right?

File: services/fxaccounts/FxAccounts.jsm [hunk 14]
```diff
@@ -392,12 +358,442 @@ function urlsafeBase64Encode(key) {
...
+   *
+   * @return Promise
+   *        Resolves with a boolean, with true indicating that we appear to
+   *        have a valid local session, or false if we need to reauthenticate
+   *        with the content server to obtain one.
+   *        Note that this doesn't check with the server - it really just tells
+   *        us if we are even able to perform that server check. To fully check
+   *        the account status, you should first call this method, and if this
+   *        returns true, you should then call sessionStatus() to check with
+   *        the server.
```
Comment (rfk) [inline services/fxaccounts/FxAccounts.jsm]: Hrm, given the advice above, I can imagine a bunch of disparate API consumers all "checking with the server" independently of each other and causing a bunch of unnecessary network traffic and server load. Perhaps we can consolidate and cache this info to ensure we only do one check every T timeperiod, to avoid slowly growing server load.

File: services/fxaccounts/FxAccounts.jsm [hunk 15]
```diff
@@ -392,12 +358,442 @@ function urlsafeBase64Encode(key) {
...
+  // XXX - new "sync" object?
+  notifyDevices(...args) {
+    return this._internal.notifyDevices(...args);
+  }
+
+  /**
+   * Resend the verification email fot the currently signed-in user.
+   *
+   */
+  resendVerificationEmail() {
```
Comment (rfk) [inline services/fxaccounts/FxAccounts.jsm]: I think we should restrict the number of places that this API gets called, e.g. maybe only from the account prefs page? Given that we're moving towards using a "type in the code from your email" style flow rather than clicking on a link, we probably want to have other UI touchpoints guide the user to a page where they can complete verification, rather than trying to direct resend the email.

Comment (rfk) [reply services/fxaccounts/FxAccounts.jsm]: (Put another way: I expect that in the future, verifying your account will involve interacting with FxA web content to enter a code, rather than just resending an email in the background.)

File: services/fxaccounts/FxAccounts.jsm [hunk 16]
```diff
@@ -392,12 +358,442 @@ function urlsafeBase64Encode(key) {
 }
 
 /**
- * The public API's constructor.
+ * The public API.
+ *
+ * XXX - TODO - *all* non-underscore stuff here should have sphinx docstrings so
+ * that docs magically appear on https://firefox-source-docs.mozilla.org/.
```
Comment (rfk) [inline services/fxaccounts/FxAccounts.jsm]: This is an impressively skinny public API, 👍 

Any thoughts on a method to call if I want to initiate signin, say from the Protection Report button?

File: services/fxaccounts/FxAccounts.jsm (hunk 15 above)
Comment (linacambridge) [reply services/fxaccounts/FxAccounts.jsm]: That lines up nicely with @mhammond's suggestion (IIUC) to have a separate "Sync" API object (like `FxAccountsKeys`), and maybe, eventually, other services?

File: services/fxaccounts/FxAccounts.jsm [hunk 17]
```diff
@@ -481,10 +761,62 @@ FxAccountsInternal.prototype = {
...
+          let scope = {};
+          ChromeUtils.import("resource://services-sync/main.js", scope);
+          return scope.Weave.Service.promiseInitialized;
+        },
+      ];
+    }
+
     this.currentTimer = null;
+    // XXX - we should probably try and get this on the "public" class, and
+    // arrange for it to be passed in to all functions here?
```
Comment (linacambridge) [inline services/fxaccounts/FxAccounts.jsm]: How do you imagine this working? Would it be up to `FxAccounts` to call `getCurrentAccountState` (or `withCurrentAccountState`?) and manage the current state? Or would `FxAccountsInternal` also hold a copy, as a "proof of state" so that it can double-check `FxAccounts` isn't holding stale state? Or is this only used for testing?

File: services/fxaccounts/FxAccountsKeys.jsm [hunk 18]
```diff
@@ -0,0 +1,315 @@
...
+      let data = await currentState.getUserAccountData();
+
+      // Sanity check that the user hasn't changed out from under us (which
+      // should be impossible given our _withCurrentAccountState, but...)
+      if (data.keyFetchToken !== keyFetchToken) {
+        throw new Error("Signed in user changed while fetching keys!");
+      }
+
+      // Next statements must be synchronous until we updateUserAccountData
+      // so that we don't risk getting into a weird state.
```
Comment (linacambridge) [inline services/fxaccounts/FxAccountsKeys.jsm]: Side note...this looks like it's no longer true, since we call `await this._deriveKeys` before `updateUserAccountData`? Is the comment stale, does this need to be within a `withCurrentAccountState`, am I totally misunderstanding this, or something else? 😸

File: services/fxaccounts/FxAccounts.jsm [hunk 19]
```diff
@@ -1687,130 +1594,60 @@ FxAccountsInternal.prototype = {
...
+        if (token) {
+          let entry = { token, server: oAuthURL };
+          // But before we do, check the cache again - if we find one now, it
+          // means someone else concurrently requested the same scope and beat
+          // us to the cache write. To be nice to the server, we revoke the one
+          // we just got and return the newly cached value.
+          let cached = currentState.getCachedToken(scope);
+          if (cached) {
+            log.debug("Detected a race for this token - revoking the new one.");
+            this._destroyOAuthToken(entry);
```
Comment (linacambridge) [inline services/fxaccounts/FxAccounts.jsm]: Nit: While you're here, does this need a `catch`?

File: services/fxaccounts/FxAccountsKeys.jsm [hunk 20]
```diff
@@ -0,0 +1,315 @@
...
+
+const { CryptoUtils } = ChromeUtils.import(
+  "resource://services-crypto/utils.js"
+);
+
+const { DERIVED_KEYS_NAMES, SCOPE_OLD_SYNC, log, logPII } = ChromeUtils.import(
+  "resource://gre/modules/FxAccountsCommon.js"
+);
+
+class FxAccountsKeys {
```
Comment (linacambridge) [inline services/fxaccounts/FxAccountsKeys.jsm]: I really like breaking out key management into a separate `FxAccountsKeys` class! :+1:

File: services/fxaccounts/FxAccountsKeys.jsm [hunk 21]
```diff
@@ -0,0 +1,315 @@
...
+
+  fetchAndUnwrapKeys(keyFetchToken) {
+    return this._fxia.withCurrentAccountState(async currentState => {
+      if (logPII) {
+        log.debug("fetchAndUnwrapKeys: token: " + keyFetchToken);
+      }
+      // Sign out if we don't have a key fetch token.
+      if (!keyFetchToken) {
+        log.warn("improper fetchAndUnwrapKeys() call: token missing");
+        await this._fxia.signOut();
```
Comment (linacambridge) [inline services/fxaccounts/FxAccountsKeys.jsm]: Hmm, `fetchAndUnwrapKeys` potentially triggering a sign out as a side effect feels surprising to me! I wonder if we can clean this up (or clean up `FxAccountsInternal`?) to return some kind of signal instead—maybe a special error—that `FxAccounts` can examine and then trigger sign out. This would help centralize state changes, and make it easier to tackle https://bugzilla.mozilla.org/show_bug.cgi?id=1256459 if we ever decide to pursue that.

OTOH, I think we're more likely to replace `FxAccounts.jsm` with our Rust library before we get to that refactor, so maybe it's not worth pushing too much on this. But, on the flip side, some of the abstractions we develop now will be helpful for building on top of the Rust client.

File: services/fxaccounts/FxAccountsProfile.jsm [hunk 22]
```diff
@@ -32,11 +32,11 @@ var FxAccountsProfile = function(options = {}) {
   this._currentFetchPromise = null;
   this._cachedAt = 0; // when we saved the cached version.
   this._isNotifying = false; // are we sending a notification?
-  this.fxa = options.fxa || fxAccounts;
+  this.fxai = options.fxai || fxAccounts._internal;
```
Comment (linacambridge) [inline services/fxaccounts/FxAccountsProfile.jsm]: Ditto, I think we can further break up FxAi with functions to manage the current state, if you're interested in pushing on that further. It seems like `FxAccountsKeys` wants a similar object.

File: services/fxaccounts/FxAccountsKeys.jsm (hunk 11 above)
Comment (mhammond) [reply services/fxaccounts/FxAccountsKeys.jsm]: yeah, filed a followup for this.

File: services/fxaccounts/FxAccountsCommands.js [hunk 23]
```diff
@@ -28,21 +28,21 @@ XPCOMUtils.defineLazyModuleGetters(this, {
 });
 
 class FxAccountsCommands {
-  constructor(fxAccounts) {
-    this._fxAccounts = fxAccounts;
-    this.sendTab = new SendTab(this, fxAccounts);
+  constructor(fxAccountsInternal) {
+    this._fxai = fxAccountsInternal;
```
[the remark this answers is not in the document]
Comment (mhammond) [reply services/fxaccounts/FxAccountsCommands.js]: good question :) Originally all of these were ending up as `this._fxa._internal` which I felt was too verbose, hence `this._fxai`. However, there's certainly a case to use `._internal` consistently, although I wonder if that might leave casual readers to imply that ever object has its own internal and not realize it's the same internal object everywhere? I'm totally open to suggestions here!

File: services/fxaccounts/FxAccountsKeys.jsm (hunk 18 above)
Comment (mhammond) [reply services/fxaccounts/FxAccountsKeys.jsm]: huh - good question :) I think things would only get into a bad state if there were multiple `updateUserAccountData` calls - that comment is in a strange place and cost more in confusion than it adds in clarify, so I just nuked it :)

File: services/fxaccounts/FxAccounts.jsm (hunk 19 above)
Comment (mhammond) [reply services/fxaccounts/FxAccounts.jsm]: I need to rebase this now that bug 561762  has landed, which I think makes this comment moot, but I'll leave this convo open to remind me to check.

File: services/fxaccounts/FxAccountsKeys.jsm (hunk 21 above)
Comment (mhammond) [reply services/fxaccounts/FxAccountsKeys.jsm]: yeah, that makes sense. I opened bug 1572313 to avoid this `signOut()`. I've been thinking a little about bug 1256459 recently and it's not clear a simple state machine would work if we consider the union of all fxa services - eg, we have sync which requires all the keys, we have notes which requires some keys, but probably not everything sync does, and we have things like monitor, which really only needs profile information (and thus doesn't even need to know about the verified state of a user). It needs more thought, and I suspect we might just try and put our rust there before we actually get back to this, but you never know :)

File: services/fxaccounts/FxAccountsProfile.jsm (hunk 22 above)
Comment (mhammond) [reply services/fxaccounts/FxAccountsProfile.jsm]: Can you please elaborate here?

File: services/fxaccounts/FxAccounts.jsm [hunk 24]
```diff
@@ -392,12 +357,365 @@ function urlsafeBase64Encode(key) {
...
+      if (!data) {
+        return null;
+      }
+      if (!FXA_ENABLED) {
+        await this.signOut();
+        return null;
+      }
+      if (this._internal.isUserEmailVerified(data)) {
+        // This is a work-around for preferences being reset (bug 1550967).
+        // Many things check this preference as a flag for "is sync configured",
```
[the remark this answers is not in the document]
Comment (mhammond) [reply services/fxaccounts/FxAccounts.jsm]: Yeah - that block is going to need to be removed for the "decoupling" work, but the more general point is good - we will need an "is this service configured" function. Ryan and I have been discussing it a bit and it's not quite obvious what this would be shaped like.

(Although really this block is for an inverted version of the problem - the flag we use to indicate the "sync service" is configured is stored in prefs, and all prefs were reset in some cases. Because Sync was the only fxa service, we could infer that "fxa signed in" must mean "sync is configured" - but in a decoupled world that would actually be impossible :( I guess that implies that any such flags we store should be in the same place the account info is stored, but that's also a bridge too far for this patch :)

File: services/fxaccounts/FxAccounts.jsm (hunk 14 above)
Comment (mhammond) [reply services/fxaccounts/FxAccounts.jsm]: Opened bug 1574051 for this

File: services/fxaccounts/FxAccounts.jsm (hunk 15 above)
Comment (mhammond) [reply services/fxaccounts/FxAccounts.jsm]: Yeah - this currently is called only from about:prefs, but we don't really have a way to enforce that currently. I think that's OK for now, and we can refactor that once we have a clearer idea about what we want to end up with.

File: services/fxaccounts/FxAccounts.jsm [hunk 25]
```diff
@@ -392,12 +357,365 @@ function urlsafeBase64Encode(key) {
...
+  async signOut(localOnly) {
+    // Note that we do not use _withCurrentAccountState here, otherwise we
+    // end up with an exception due to the user signing out before the call is
+    // complete - but that's the entire point of this method :)
+    return this._internal.signOut(localOnly);
+  }
+
+  // XXX - we should consider killing this - the only reason it is public is
+  // so that sync can change it when it notices the device name being changed,
+  // and that could probably be replaced with a pref observer.
```
[the remark this answers is not in the document]
Comment (mhammond) [reply services/fxaccounts/FxAccounts.jsm]: yes, but in my current structure, our push code is free to call internal functions.

File: services/sync/modules/browserid_identity.js (hunk 8 above)
Comment (mhammond) [reply services/sync/modules/browserid_identity.js]: As per Ryan's previous comment in this file, let's leave this module as a bit of a special case in the short term and clean things up later when we move sync to oauth tokens.

**Comment by mhammond** [comment]
I think this is ready for review. I'll move it to phabricator soon, but feel free to have a look here.
